### ISSUE TITLE
Periscope should clean itself up after completing

### ISSUE THREAD

**Issue by davidkydd** [Azure/aks-periscope#23]
Currently periscope requires manual cleanup from the cluster after it completes uploading logs, otherwise it leaves behind namespace, pods etc. If not cleaned up it is both wasteful of cluster resources (e.g. counts towards pod limits) and will prevent periscope from being deployed and executing a second time.

This manual cleanup introduces a burden on all periscope clients.

It should be possible to instrument periscope to clean up itself once complete, by giving it delete namespace and deployment permissions for the aks-periscope namespace via rbac.

**Comment by davidkydd** [comment]
This needs something like a daemonJob and isn't straightforward if done in the periscope executable itself. Due to daemonSet only supporting --restart-always, we would need to do something like have pods synchronize before the call to delete (e.g. how can the last running pod know the other pods have completed, and that it should call delete). Perhaps one of the ideas here https://github.com/kubernetes/kubernetes/issues/64623 or this implementation: https://github.com/AmitKumarDas/metac/tree/master/examples/daemonjob

**Comment by peterbom** [comment]
I think this breaks down into a number of alternative considerations:
1. Should we alter the existing behaviour so that pods within the DaemonSet get cleaned up after use?
2. Should we be doing anything to make it more obvious to users what is being deployed, and how to clean it up themselves?
3. Should we be using some other Kubernetes resource than a DaemonSet?

For (1), I think we can say 'no'. The standard resource types have a set of expectations about how they will behave, and the nature of a DaemonSet is that it's expected to stay running. As a user, I'd find it surprising if I defined and deployed a DaemonSet that launched pods on every node, and then sometime later I found they were no longer there. Given Kubernetes' goal-seeking approach I think that could only happen if the resource spec changed itself, which would be pretty confusing. It would also break the idempotence of successive identical deployments.

For (2), I think this is only a consideration if the deployment is abstracted away from the user (e.g. via the VS Code extension). If the user is deploying resources to their cluster explicitly (via kubectl commands, or Helm/Porter) it will be trivial for them to clean up those resources themselves.

For (3), this warrants further discussion. If we could find a trustworthy, reliable, well-tested resource type that's better aligned with the job-like nature of Periscope, that is something we should spend some time considering. I'll create a separate discussion item for this.
